actionpack/lib/action_dispatch/routing/mapper.rb

```diff
@@ -1168,13 +1168,27 @@ module Resources
         CANONICAL_ACTIONS = %w(index create new show update destroy)
 
         class Resource # :nodoc:
+          class << self
+            def default_actions(api_only)
+              if api_only
+                [:index, :create, :show, :update, :destroy]
+              else
+                [:index, :create, :new, :show, :update, :destroy, :edit]
+              end
+            end
+          end
+
           attr_reader :controller, :path, :param
 
           def initialize(entities, api_only, shallow, options = {})
             if options[:param].to_s.include?(":")
               raise ArgumentError, ":param option can't contain colons"
             end
 
+            if (invalid_actions = (options.values_at(:only, :except).flatten.compact - default_actions)).any?
```

Comment:
`(options.values_at(:only, :except).flatten.compact - default_actions)` could be extracted to an `invalid_actions` method.